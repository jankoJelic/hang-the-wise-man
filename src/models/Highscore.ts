type Highscore = {
  id: string;
  length: number;
  userName: string;
  duration: number;
  quoteId: string;
  errors: number;
  uniqueCharacters: number;
};

export default Highscore;
