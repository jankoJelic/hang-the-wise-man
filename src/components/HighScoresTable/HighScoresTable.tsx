import useWindowSize from "hooks/useWindowsSize";
import Highscore from "models/Highscore";
import calculateRealScore from "util/calculateRealScore";

const HighScoresTable = ({ data }: { data: Highscore[] }) => {
  const { width } = useWindowSize();

  const score = (errors: number) => parseFloat((100 / (1 + errors)).toFixed(2));

  const dataWithScore = data.map((hs) => {
    return {
      ...hs,
      score: score(hs.errors),
      realScore: calculateRealScore({
        quoteLength: hs.length,
        uniqueLetters: hs.uniqueCharacters,
        errors: hs.errors,
        solvingTime: hs.duration,
      }),
    };
  });

  const dataByScore = dataWithScore.sort((a, b) => b.score - a.score);

  return (
    <table
      style={{ width: width * 0.9 }}
      className="bg-white rounded-xl table-auto p-4 mb-8 overflow-hidden shadow-md"
    >
      <thead className="h-8 border-b-2 border-mainText">
        <tr>
          <th>Username</th>
          <th>Length</th>
          <th>Unique Characters</th>
          <th>Duration</th>
          <th className="text-mainAction">Score</th>
          <th>Real Score</th>
        </tr>
      </thead>
      <tbody>
        {dataByScore.map((hs, index) => {
          console.log(hs);
          return (
            <tr
              key={hs.id}
              className={`text-center py-1 roundex-xl ${
                index % 2 ? "bg-gray-100" : ""
              }`}
            >
              <td className="">{hs.userName}</td>
              <td>{hs.length}</td>
              <td>{hs.uniqueCharacters}</td>
              <td>{hs.duration / 1000}</td>
              <td className="font-semibold text-mainAction">{hs.score}</td>
              <td>{hs.realScore.toFixed(5)}</td>
            </tr>
          );
        })}
      </tbody>
    </table>
  );
};

export default HighScoresTable;
