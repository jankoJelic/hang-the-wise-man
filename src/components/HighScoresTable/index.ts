import HighScoresTable from "./HighScoresTable";

export default HighScoresTable;
