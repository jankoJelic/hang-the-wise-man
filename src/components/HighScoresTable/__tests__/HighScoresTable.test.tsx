import HighScoresTable from "../HighScoresTable";
import renderer from "react-test-renderer";
import { render, screen } from "@testing-library/react";
import Highscore from "models/Highscore";

const mockData: Highscore[] = [
  {
    id: "1",
    length: 20,
    userName: "janko",
    duration: 100000,
    quoteId: "e2ni",
    errors: 10,
    uniqueCharacters: 17,
  },
  {
    id: "2",
    length: 22,
    userName: "jakov",
    duration: 222222,
    quoteId: "e22ni",
    errors: 9,
    uniqueCharacters: 14,
  },
  {
    id: "3",
    length: 99,
    userName: "user3",
    duration: 88888,
    quoteId: "e12ni",
    errors: 3,
    uniqueCharacters: 20,
  },
];

const renderTable = () => render(<HighScoresTable data={mockData} />);

it("renders correctly", () => {
  const tree = renderer.create(<HighScoresTable data={mockData} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("has 6 columns", () => {
  renderTable();

  expect(screen.getAllByRole("columnheader")).toHaveLength(6);
});

it("its body contains one row for each item in data", () => {
  renderTable();

  expect(screen.getAllByRole("cell")).toHaveLength(6 * mockData.length);
});
