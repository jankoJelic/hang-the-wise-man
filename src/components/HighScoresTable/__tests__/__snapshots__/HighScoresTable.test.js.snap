// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="bg-white rounded-xl table-auto p-4 mb-8 overflow-hidden shadow-md"
  style={
    Object {
      "width": 0,
    }
  }
>
  <thead
    className="h-8 border-b-2 border-mainText"
  >
    <tr>
      <th>
        Username
      </th>
      <th>
        Length
      </th>
      <th>
        Unique Characters
      </th>
      <th>
        Duration
      </th>
      <th
        className="text-mainAction"
      >
        Score
      </th>
      <th>
        Real Score
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      className="text-center py-1 roundex-xl "
    >
      <td
        className=""
      >
        user3
      </td>
      <td>
        99
      </td>
      <td>
        20
      </td>
      <td>
        88.888
      </td>
      <td
        className="font-semibold text-mainAction"
      >
        25
      </td>
      <td>
        25.02099
      </td>
    </tr>
    <tr
      className="text-center py-1 roundex-xl bg-gray-100"
    >
      <td
        className=""
      >
        jakov
      </td>
      <td>
        22
      </td>
      <td>
        14
      </td>
      <td>
        222.222
      </td>
      <td
        className="font-semibold text-mainAction"
      >
        10
      </td>
      <td>
        10.01422
      </td>
    </tr>
    <tr
      className="text-center py-1 roundex-xl "
    >
      <td
        className=""
      >
        janko
      </td>
      <td>
        20
      </td>
      <td>
        17
      </td>
      <td>
        100
      </td>
      <td
        className="font-semibold text-mainAction"
      >
        9.09
      </td>
      <td>
        9.10811
      </td>
    </tr>
  </tbody>
</table>
`;
